// @fluidic/sdk/dist/evm.js
import { FLUIDIC_EVM_CHAIN_ID } from "./constants.js";
export { FLUIDIC_EVM_CHAIN_ID };
/** JSON-RPC provider that targets the Fluidic EVM gateway. */
export class FluidicEvmProvider {
    client;
    constructor(client) {
        this.client = client;
    }
    async rpc(method, params) {
        const query = this.client["minTickQuery"]() ?? "";
        const url = `${this.client.apiUrl}/rpc${query}`;
        const res = await fetch(url, {
            method: "POST",
            headers: { "Content-Type": "application/json" },
            body: JSON.stringify({
                jsonrpc: "2.0",
                id: 1,
                method,
                params,
            }),
        });
        if (!res.ok) {
            throw new Error(`EVM RPC HTTP ${res.status}`);
        }
        const json = (await res.json());
        if (json.error) {
            throw new Error(`EVM RPC ${json.error.code}: ${json.error.message}`);
        }
        return json.result;
    }
    async blockNumber() {
        const result = await this.rpc("eth_blockNumber", []);
        return BigInt(String(result));
    }
    async chainId() {
        const result = await this.rpc("eth_chainId", []);
        return BigInt(String(result));
    }
    async getBalance(address) {
        const result = await this.rpc("eth_getBalance", [address, "latest"]);
        return BigInt(String(result));
    }
    async getTransactionCount(address) {
        const result = await this.rpc("eth_getTransactionCount", [address, "latest"]);
        return BigInt(String(result));
    }
    async sendRawTransaction(signedTx) {
        const result = await this.rpc("eth_sendRawTransaction", [signedTx]);
        return String(result);
    }
    async getTransactionReceipt(hash) {
        return this.rpc("eth_getTransactionReceipt", [hash]);
    }
    async waitForReceipt(hash, options = {}) {
        const pollMs = options.pollMs ?? 500;
        const deadline = Date.now() + (options.timeoutMs ?? 60_000);
        while (Date.now() < deadline) {
            const receipt = await this.getTransactionReceipt(hash);
            if (receipt !== null && receipt !== undefined)
                return receipt;
            await new Promise((r) => setTimeout(r, pollMs));
        }
        throw new Error(`Timeout waiting for EVM receipt ${hash}`);
    }
}
/** Poll an EVM transaction receipt until it is no longer null. */
export async function waitForEvmReceipt(provider, hash, options = {}) {
    return provider.waitForReceipt(hash, options);
}
