// @fluidic/sdk/dist/shifts.js
import { bytesToHex, concatBytes, coordinateToBytes, ensureHex32, hashBlake3, hexToBytes, i128ToLeBytes, nowNs, u128ToLeBytes, u64ToLeBytes, } from "./utils.js";
export const DEFAULT_DEX_DOMAIN = "4445585f574156455f5553444300000000000000000000000000000000000000";
function vectorClockBytes(vc) {
    const entries = Object.entries(vc.entries)
        .map(([node, time]) => ({ node: ensureHex32(node), time }))
        .sort((a, b) => (a.node > b.node ? 1 : -1));
    const parts = [];
    for (const { node, time } of entries) {
        parts.push(hexToBytes(node));
        parts.push(u64ToLeBytes(time));
    }
    return concatBytes(...parts);
}
export function buildCommutativeShift(params) {
    const domain = params.domain ?? DEFAULT_DEX_DOMAIN;
    const nonce = params.nonce ?? 0n;
    const timestampNs = params.timestampNs ?? nowNs();
    const signingBytes = concatBytes(new TextEncoder().encode("FLUIDIC:COMMUTATIVE:v2"), hexToBytes(ensureHex32(domain)), coordinateToBytes(params.coordinate), i128ToLeBytes(params.delta), hexToBytes(ensureHex32(params.poolId)), u64ToLeBytes(nonce), u64ToLeBytes(timestampNs));
    const signature = params.signer.sign(signingBytes);
    return {
        domain,
        coordinate: params.coordinate,
        delta: params.delta,
        pool_id: params.poolId,
        nonce,
        timestamp_ns: timestampNs,
        signature: bytesToHex(signature),
    };
}
export function buildStatefulShift(params) {
    const from = params.signer.accountId;
    const domain = params.domain ?? DEFAULT_DEX_DOMAIN;
    const predecessors = params.predecessors ?? [];
    const nonce = params.nonce ?? 0n;
    const timestampNs = params.timestampNs ?? nowNs();
    const signingBytes = concatBytes(new TextEncoder().encode("FLUIDIC:STATEFUL:v2"), hexToBytes(ensureHex32(domain)), hexToBytes(ensureHex32(from)), hexToBytes(ensureHex32(params.to)), u128ToLeBytes(params.amount), vectorClockBytes(params.vectorClock), concatBytes(...predecessors.map((h) => hexToBytes(ensureHex32(h)))), u64ToLeBytes(nonce), u64ToLeBytes(timestampNs));
    const signature = params.signer.sign(signingBytes);
    return {
        domain,
        from,
        to: params.to,
        amount: params.amount,
        vector_clock: params.vectorClock,
        predecessors,
        nonce,
        timestamp_ns: timestampNs,
        signature: bytesToHex(signature),
    };
}
export function buildRegistrationShift(params) {
    return {
        account: params.signer.accountId,
        public_key: bytesToHex(params.signer.publicKey),
        wave_account: params.signer.waveAccount,
        usdc_account: params.signer.usdcAccount,
        nonce: params.nonce ?? 0n,
        timestamp_ns: params.timestampNs ?? nowNs(),
    };
}
export function buildStakeShift(params) {
    const operator = params.signer.accountId;
    const publicKey = bytesToHex(params.signer.publicKey);
    const nonce = params.nonce ?? 0n;
    const timestampNs = params.timestampNs ?? nowNs();
    const signingBytes = concatBytes(new TextEncoder().encode("FLUIDIC:STAKE:v1"), hexToBytes(ensureHex32(operator)), hexToBytes(ensureHex32(publicKey)), u128ToLeBytes(params.amount), u64ToLeBytes(nonce), u64ToLeBytes(timestampNs));
    const signature = params.signer.sign(signingBytes);
    return {
        operator,
        public_key: publicKey,
        amount: params.amount,
        nonce,
        timestamp_ns: timestampNs,
        signature: bytesToHex(signature),
    };
}
export function hashStatefulShift(shift) {
    const signingBytes = concatBytes(new TextEncoder().encode("FLUIDIC:STATEFUL:v2"), hexToBytes(ensureHex32(shift.domain)), hexToBytes(ensureHex32(shift.from)), hexToBytes(ensureHex32(shift.to)), u128ToLeBytes(shift.amount), vectorClockBytes(shift.vector_clock), concatBytes(...shift.predecessors.map((h) => hexToBytes(ensureHex32(h)))), u64ToLeBytes(shift.nonce), u64ToLeBytes(shift.timestamp_ns));
    const sig = hexToBytes(shift.signature);
    return bytesToHex(hashBlake3(concatBytes(signingBytes, sig)));
}
export function hashCommutativeShift(shift) {
    const signingBytes = concatBytes(new TextEncoder().encode("FLUIDIC:COMMUTATIVE:v2"), hexToBytes(ensureHex32(shift.domain)), coordinateToBytes(shift.coordinate), i128ToLeBytes(shift.delta), hexToBytes(ensureHex32(shift.pool_id)), u64ToLeBytes(shift.nonce), u64ToLeBytes(shift.timestamp_ns));
    const sig = hexToBytes(shift.signature);
    return bytesToHex(hashBlake3(concatBytes(signingBytes, sig)));
}
